subsys/net/lib/lwm2m/lwm2m_obj_connmon.c

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright (c) 2019 Foundries.io
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_obj_conn_mon
+#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <string.h>
+#include <init.h>
+#include <net/net_if.h>
+#include <net/net_ip.h>
+
+#include "lwm2m_object.h"
+#include "lwm2m_engine.h"
+
+/* Connectivity Monitoring resource IDs */
+#define CONNMON_NETWORK_BEARER_ID		0
+#define CONNMON_AVAIL_NETWORK_BEARER_ID		1
+#define CONNMON_RADIO_SIGNAL_STRENGTH		2
+#define CONNMON_LINK_QUALITY			3
+#define CONNMON_IP_ADDRESSES			4
+#define CONNMON_ROUTER_IP_ADDRESSES		5
+#define CONNMON_LINK_UTILIZATION		6
+#define CONNMON_APN				7
+#define CONNMON_CELLID				8
+#define CONNMON_SMNC				9
+#define CONNMON_SMCC				10
+
+#define CONNMON_MAX_ID				11
+
+#define CONNMON_STRING_SHORT			8
+
+#if defined(CONFIG_LWM2M_CONN_MON_BEARER_MAX)
+#define CONNMON_AVAIL_BEARER_MAX	CONFIG_LWM2M_CONN_MON_BEARER_MAX
+#else
+#define CONNMON_AVAIL_BEARER_MAX	5
+#endif
+
+#if defined(CONFIG_LWM2M_CONN_MON_APN_MAX)
+#define CONNMON_APN_MAX			CONFIG_LWM2M_CONN_MON_APN_MAX
+#else
+#define CONNMON_APN_MAX			5
+#endif
+
+#if defined(CONFIG_NET_IF_MAX_IPV6_COUNT) \
+	&& defined(CONFIG_NET_IF_MAX_IPV6_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		(CONFIG_NET_IF_MAX_IPV6_COUNT + \
+					CONFIG_NET_IF_MAX_IPV4_COUNT)
+#elif defined(CONFIG_NET_IF_MAX_IPV6_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		CONFIG_NET_IF_MAX_IPV6_COUNT
+#elif defined(CONFIG_NET_IF_MAX_IPV4_COUNT)
+#define CONNMON_IP_ADDRESS_MAX		CONFIG_NET_IF_MAX_IPV4_COUNT
+#else
+#define CONNMON_IP_ADDRESS_MAX		1
+#endif
+
+#if defined(CONFIG_NET_MAX_ROUTERS)
+#define CONNMON_ROUTER_IP_ADDRESS_MAX	CONFIG_NET_MAX_ROUTERS
+#else
+#define CONNMON_ROUTER_IP_ADDRESS_MAX	1
+#endif
+
+/* resource state variables */
+static s8_t net_bearer;
+static s8_t avail_net_bearer[CONNMON_AVAIL_BEARER_MAX];
+static s8_t rss;
+static u8_t link_quality;
+static char ip_address[CONNMON_IP_ADDRESS_MAX][INET6_ADDRSTRLEN];
```

Comment:
These IP address strings use quite a lot of memory as the IPv6 string is 46 bytes long. Would it make sense to store the binary value instead and just convert to/from string when needed?

Reply:
My original patches used `struct net_addr` for storage.  I can revert back to that or some equivalent.  In my head implementing the IPs as strings seemed easier for the user, so I changed.

It also exposed a much deeper issue I'm having with resource instances atm.  It's just too hard for the user to add / delete these instances in any sane manner.  Each of these multi-instance resource is far too hard to work with at the moment.  I have another patch set I'm working on to make creating / deleting and setting the multi-instance resources basically the same as any other resource.

Here's an example from the lwm2m sample where we set the data buffer for the `Manufacturer` field:
```
	lwm2m_engine_set_res_data("3/0/0", CLIENT_MANUFACTURER,
				  sizeof(CLIENT_MANUFACTURER),
				  LWM2M_RES_DATA_FLAG_RO);
```
We access the resource via the established "3/0/0" URI (object id, object instance id and resource id)

For us to set the first 2 instances of the Connectivity Monitoring IP addresses we *should* be able to use something like:
```
	lwm2m_engine_create_res_inst("4/0/4/0");
	lwm2m_engine_set_res_data("4/0/4/0", const_string, length of cont_string,
				  LWM2M_RES_DATA_FLAG_RO);
	lwm2m_engine_create_res_inst("4/0/4/1");
	lwm2m_engine_set_res_data("4/0/4/1", const_string2, length of cont_string2,
				  LWM2M_RES_DATA_FLAG_RO);
```
Not only does this remove the need for pre-allocating the space for these IP addresses, but also allows for non-sequential resource instance id's etc.  (It's up to the user).   The user could also setup a pre-read callback where it takes the current IP address and converts it to a string on the fly and hands it back whenever it's read.  Etc.  Endless more opportunities.

Instead, because of the way the resource instances are implemented atm, we use these crazy 1-off functions like:
`lwm2m_connmon_add_bearer()`
`lwm2m_connmon_remove_bearer()`
`lwm2m_connmon_add_ip_address()`
and we have to pre-allocate all of the resource data during the object instance creation.

Reply:
> Not only does this remove the need for pre-allocating the space for these IP addresses, but also allows for non-sequential resource instance id's etc. (It's up to the user).

Indeed this sounds much better solution. We can leave the strings for now if in the future we do not really need them anyway.